crypto/zk/artifacts/artifacts.go

```diff
@@ -0,0 +1,179 @@
+package artifacts
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+)
+
+/*
+# Directories structure
+=======================
+Path: circuitname/network/parameters/file.extension
+- Circuit Name: the circuit name. Example: `zkcensusproof`
+- Network: The network name. Example: `main, dev, stage`
+- Circuit Parameters: numbers in decimal separated by underscore. Example: `1024`, `1024_3_500`
+- Files:
+	- circuit.zkey: contains the provingKey & verificationKey (big size, linear with circuit
+	size) [needed by the user]
+	- verificationKey.json: extracted from circuit.zkey (small size, constant size) [needed
+	by the vochain]
+	- witness.wasm: (medium size, linear with circuit size) [needed by the user]
+- Examples:
+	- zkcensusproof/stage/1024/circuit.zkey
+	- zkcensusproof/dev/128/witness.wasm
+
+# Files hash
+=============
+Each file is hashed with sha256, which output is publicly known. The hash
+values of each file are hardcoded in the VochainGenesis.
+
+*/
+
+const (
+	// FilenameWitness defines the name of the file of the WitnessCalculator
+	FilenameWitness string = "witness.wasm"
+	// FilenameZKey defines the name of the file of the circom ZKey
+	FilenameZKey string = "circuit.zkey"
+	// FilenameVK defines the name of the VerificationKey.json
+	FilenameVK string = "verificationKey.json"
+)
+
+// CircuitsConfig defines the configuration of the files to be downloaded
+type CircuitsConfig struct {
+	// URL defines the URL from where to download the files
+	URL string
+	// CircuitsPath defines the path from where the files are downloaded
+	CircuitsPath string
+	// LocalDir defines in which directory will be the files
+	// downloaded, under that directory it will follow the CircuitsPath
+	// directories structure
+	LocalDir string
+
+	// WitnessHash contains the expected hash for the file filenameWitness
+	WitnessHash []byte // witness.wasm
+	// ZKeyHash contains the expected hash for the file filenameZKey
+	ZKeyHash []byte // circuit.zkey
+	// VKHash contains the expected hash for the file filenameVK
+	VKHash []byte // verificationKey.json
+}
+
+// DownloadCircuitFiles will download the circuits in the specified path,
+// checking the expected sha256 hash of each file. If the files already exist
+// in the path, it will not download them again but it will check the hash of
+// the existing files.
+func DownloadCircuitFiles(ctx context.Context, c CircuitsConfig) error {
+	// check if files already exist, if files already exist, check hashes
+	err := checkHashes(c)
+	if err == nil {
+		// files already exist, and match the expected hashes
+		return nil
+	}
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	// at this point, or files do not exist, or exist with a not expected
+	// hash
+	return downloadFiles(ctx, c)
+}
+
+func checkHashes(c CircuitsConfig) error {
+	err := checkHash(c.LocalDir+c.CircuitsPath+FilenameWitness, c.WitnessHash)
```

Comment:
you should be using filepath.Join for paths on disk; otherwise this is not very portable outside unix-like systems